// cozy-notifications-helper/lib/main.js
// Generated by CoffeeScript 1.6.3
(function() {
  var NotificationManager, NotificationsHelper;

  NotificationManager = require('./notification-manager');

  module.exports = NotificationsHelper = (function() {
    function NotificationsHelper(app) {
      this.app = app;
    }

    NotificationsHelper.prototype.createTemporary = function(params, callback) {
      if (callback == null) {
        callback = function() {};
      }
      return NotificationManager.manage(params, 'temporary', callback);
    };

    NotificationsHelper.prototype.createOrUpdatePersistent = function(ref, params, callback) {
      var _ref;
      if (callback == null) {
        callback = function() {};
      }
      params.ref = ref;
      if ((params != null ? (_ref = params.resource) != null ? _ref.app : void 0 : void 0) != null) {
        params.app = this.app;
      } else if (this.app != null) {
        params.app = params.resource.app;
      }
      return NotificationManager.manage(params, 'persistent', callback);
    };

    NotificationsHelper.prototype.destroy = function(ref, callback) {
      var params;
      if (callback == null) {
        callback = function() {};
      }
      params = {
        ref: ref,
        app: this.app != null ? this.app : void 0
      };
      return NotificationManager.destroy(params, callback);
    };

    return NotificationsHelper;

  })();

}).call(this);
